File: 300-longest-increasing-subsequence/longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        # if not nums:
        #     return 0
        
        # n = len(nums)
        # dp = [1] * n

        # for i in range(1, n):
        #     for j in range(i):
        #         if nums[i] > nums[j]:
        #             dp[i] = max(dp[i], dp[j] + 1)

        # return max(dp)

        monostack = []
        for num in nums: 
            if not monostack or monostack[-1] < num:
                monostack.append(num)
            else:
                idx = bisect.bisect_left(monostack, num)
                monostack[idx] = num
        return len(monostack)

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.ts
function lengthOfLIS(nums: number[]): number {
    if (!nums.length) return 0;
    
    // tails[i] represents the smallest tail of all increasing subsequences of length i+1
    const tails: number[] = [];
    
    for (const num of nums) {
        // Find the position to insert num using binary search
        let left = 0;
        let right = tails.length;
        
        while (left < right) {
            const mid = Math.floor((left + right) / 2);
            if (tails[mid] < num) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        
        // If at end, extend the longest sequence
        if (left === tails.length) {
            tails.push(num);
        }
        // Otherwise replace the smallest number >= num
        else {
            tails[left] = num;
        }
    }
    
    return tails.length;
};